_posts/2023-09-15-kafka-producer-client-essentials.md

```diff
@@ -0,0 +1,372 @@
+    > You can also use the `retries` configuration property to control how many times the producer will retry sending a message before giving up. 
+    > This setting affects how many times the `retriable` method may return `true` during a message send error. 
+    > For more information, see the [Strimzi documentation on ordered delivery](https://strimzi.io/docs/operators/latest/deploying#ordered_delivery).
+
+**`onCompletion` method**
+
+- Confirms successful message transmission and displays information about the message sent, including the topic, partition, and offset.
+- Prints an error message on exception. Appropriate action is taken based on whether it's a retriable or non-retriable error. If the error is retriable, the message sending process continues. If the error is non-retriable, a stack trace is printed and the producer is terminated.
+
+With the imported libraries, our constants, and these configuration properties and methods, the producer application can do all we set out to do.
+
+**Example producer application**
+```java
+import java.util.Properties;
+import java.util.Random;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.apache.kafka.clients.producer.Callback;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.errors.RetriableException;
+import org.apache.kafka.common.serialization.ByteArraySerializer;
+import org.apache.kafka.common.serialization.LongSerializer;
+
+public class Producer implements Callback {
+    // Constants for configuration
+    private static final Random RND = new Random(0);
+    private static final String BOOTSTRAP_SERVERS = "localhost:9092";
+    private static final String TOPIC_NAME = "my-topic";
+    private static final long NUM_MESSAGES = 100;
+    private static final int MESSAGE_SIZE_BYTES = 100;
+    private static final long PROCESSING_DELAY_MS = 0L;
+
+    protected AtomicLong messageCount = new AtomicLong(0);
+
+    public static void main(String[] args) {
+        new Producer().run();
+    }
+
+    public void run() {
+        System.out.println("Running producer");
+
+        // Create a Kafka producer instance
+        // This producer sends messages to the Kafka topic asynchronously
+        try (var producer = createKafkaProducer()) {
+            // Generate a random byte array as the message payload
+            byte[] value = randomBytes(MESSAGE_SIZE_BYTES);
+            while (messageCount.get() < NUM_MESSAGES) {
+                sleep(PROCESSING_DELAY_MS);                
+                // Send a message to the Kafka topic, specifying topic name, message count, and message value
+                producer.send(new ProducerRecord<>(TOPIC_NAME, messageCount.get(), value), this);
```

Comment:
This is a little confusing. It looks like the `ProducerRecord` takes a `messageCount` parameter, but we're actually taking `messageCount` as the message key. Could we explain it here?

Reply:
Changed to:  `// Send a message to the Kafka topic, specifying topic name, message key, and message value`